### ISSUE TITLE
Why the graph pattern of _convert_attention_mask is changed in paddle2.5?

### ISSUE THREAD

**Issue by Wong4j** [PaddlePaddle/Paddle#55829]
### 请提出你的问题 Please ask your question

We have integrated cuDNN fused kernel into Paddle 2.4 to fuse DotProductAttention. We designed a build strategy, which detect the subgraph of attention and replace it with fused kernel in AMP training.
But in paddle 2.5, the graph pattern detection failed. Because the graph pattern is different from paddle 2.4. One of the main difference is [_convert_attention_mask](https://github.com/PaddlePaddle/Paddle/blob/60e37d178a5a26a6e57e1828a14164854800bfd7/python/paddle/nn/layer/transformer.py#L84), which calculates `mask = (paddle.cast(mask, 'float16')-1.0) * 1e9`.

In paddle 2.4, the graph pattern is `mask_int32 -> cast op (to fp32) -> scale (-1.0) -> scale (*1e9) -> cast op (to fp16) -> mask_fp16` as following:
```
{Out=['cast_0.tmp_0']} = cast(inputs={X=['attn_mask']}, in_dtype = 2, op_device = , op_namescope = /, op_role = 0, op_role_var = [], out_dtype = 5, use_mkldnn = False, with_quant_attr = False)
{Out=['tmp_1']} = scale(inputs={ScaleTensor=[], X=['cast_0.tmp_0']}, bias = -1.0, bias_after_scale = True, op_device = , op_namescope = /, op_role = 0, op_role_var = [], scale = 1.0, with_quant_attr = False)
{Out=['tmp_2']} = scale(inputs={ScaleTensor=[], X=['tmp_1']}, bias = 0.0, bias_after_scale = True, op_device = , op_namescope = /, op_role = 0, 
op_role_var = [], scale = 1000000000.0, with_quant_attr = False)
{Out=['tmp_2.cast_fp16']} = cast(inputs={X=['tmp_2']}, in_dtype = 5, op_device = , op_namescope = /, op_role = 0, op_role_var = [], out_dtype = 
4, use_mkldnn = False, with_quant_attr = False)
```
In paddle 2.5, the graph pattern is `mask_int32 -> cast op (to fp32) -> cast op (to fp16) -> scale op (-1.0) -> scale op (*1e9) -> mask_fp16` as following:
```
{Out=['cast_0.tmp_0']} = cast(inputs={X=['attn_mask']}, in_dtype = 2, op_device = , op_namescope = /, op_role = 0, op_role_var = [], out_dtype = 5, use_mkldnn = False, with_quant_attr = False)
{Out=['cast_0.tmp_0.cast_fp16']} = cast(inputs={X=['cast_0.tmp_0']}, in_dtype = 5, op_device = , op_namescope = /, op_role = 0, op_role_var = [], out_dtype = 4, use_mkldnn = False, with_quant_attr = False)
{Out=['tmp_1']} = scale(inputs={ScaleTensor=[], X=['cast_0.tmp_0.cast_fp16']}, bias = -1.0, bias_after_scale = True, op_device = , op_namescope 
= /, op_role = 0, op_role_var = [], scale = 1.0, with_quant_attr = False)
{Out=['tmp_2']} = scale(inputs={ScaleTensor=[], X=['tmp_1']}, bias = 0.0, bias_after_scale = True, op_device = , op_namescope = /, op_role = 0, 
op_role_var = [], scale = 1000000000.0, with_quant_attr = False)
```

- Why does this difference occur with the same [_convert_attention_mask](https://github.com/PaddlePaddle/Paddle/blob/60e37d178a5a26a6e57e1828a14164854800bfd7/python/paddle/nn/layer/transformer.py#L84) code?
- Why can't the int32 mask be converted to fp16 directly? 

To reproduce, run this code in paddle 2.4 & 2.5 and compare the `main_program`:
```
import paddle
import numpy as np

paddle.enable_static()

hidden_dim = 1024
seqlen = 128


class mha(paddle.nn.Layer):
    def __init__(
        self,
        hidden,
        num_heads,
        dropout=0.0,
    ):
        super().__init__()
        self.mha_layer = paddle.nn.MultiHeadAttention(hidden, num_heads, dropout=dropout)

    def forward(self, q, mask):
        out = self.mha_layer(q, q, q, attn_mask=mask)
        loss = paddle.mean(out)
        return loss

main_prog = paddle.static.Program()
startup_prog = paddle.static.Program()

with paddle.static.program_guard(main_prog, startup_prog):
    q_input = paddle.static.data(
        name="q_input",
        shape=[-1, -1, hidden_dim],
        dtype='float32',
    )
    attn_mask = paddle.static.data(
        name="attn_mask",
        shape=[-1, 1, 1, seqlen],
        dtype='int32',
    )
    model = mha(hidden=hidden_dim, num_heads=16)

    loss = model(q_input, attn_mask,)
    opt = paddle.optimizer.SGD(learning_rate=0.1)

    opt = paddle.static.amp.decorate(
        optimizer=opt,
        init_loss_scaling=128.0,
        use_dynamic_loss_scaling=True,
    )
    opt.minimize(loss)

_q_input = np.random.normal(
    loc=0.0,
    scale=1.0,
    size=(2, seqlen, hidden_dim),
).astype(np.float32)

_attn_mask = np.ones(
    shape=(2, 1, 1, seqlen),
    dtype=np.int32,
)

feed = {
    "q_input": _q_input,
    "attn_mask": _attn_mask,
}


exe = paddle.static.Executor(paddle.CUDAPlace(0))
print(main_prog)

exe.run(startup_prog)
exe.run(main_prog, feed)
```

**Comment by minghaoBD** [comment]
Could you provide the links to 2.4 and 2.5? I didn't see any differences in _convert_attention_mask:

2.4: https://github.com/PaddlePaddle/Paddle/blame/release/2.4/python/paddle/nn/layer/transformer.py#L80
2.5: https://github.com/PaddlePaddle/Paddle/blame/release/2.5/python/paddle/nn/layer/transformer.py#L84

**Comment by Wong4j** [comment]
@minghaoBD Yeah, I cannot find any difference in _convert_attention_mask between 2.4 & 2.5 either. 
I just run BERT pretraning in paddle 2.4 & 2.5, and found the graph is different. 

You can run the repro code I provide above in 2.4 & 2.5 and compare the main_program.
